proposals/0035-linalg-matrix.md

```diff
@@ -39,126 +39,106 @@ Matrix Operations](/proposals/0031-hlsl-vector-matrix-operations.md).
-  Cos = 4,
-  Tan = 5,
-};
-
 template <MatrixComponentType ComponentTy, uint M, uint N, MatrixUse Use,
           MatrixScope Scope>
 class Matrix {
-  using ElementType = __detail::ComponentTypeTraits<ComponentTy>::Type;
+  using ElementType = typename __detail::ComponentTypeTraits<ComponentTy>::Type;
 
   template <MatrixComponentType NewCompTy, MatrixUse NewUse = Use>
   Matrix<NewCompTy, M, N, NewUse, Scope> cast();
 
   // Element-wise operations
   template <typename T>
-    requires ArithmeticScalar<T>
-  Matrix operator+=(T);
+  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value, Matrix>::type
+  operator+=(T);
   template <typename T>
-    requires ArithmeticScalar<T>
-  Matrix operator-=(T);
+  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value, Matrix>::type
+  operator-=(T);
   template <typename T>
-    requires ArithmeticScalar<T>
-  Matrix operator*=(T);
+  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value, Matrix>::type
+  operator*=(T);
   template <typename T>
-    requires ArithmeticScalar<T>
-  Matrix operator/=(T);
+  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value, Matrix>::type
+  operator/=(T);
 
   // Apply a unary operation to each element.
-  template<UnaryOperation Op>
-  Matrix ApplyUnaryOperation();
+  template <UnaryOperation Op> Matrix ApplyUnaryOperation();
 
   template <typename T>
-    requires ArithmeticScalar<T>
-  static Matrix Splat(T Val);
-  static Matrix
-  Load(ByteAddressBuffer Res, uint StartOffset, uint Stride, bool ColMajor,
-       uint Align = sizeof(ElementType));
-  static Matrix
-  Load(RWByteAddressBuffer Res, uint StartOffset, uint Stride, bool ColMajor,
-       uint Align = sizeof(ElementType));
+  static typename hlsl::enable_if<hlsl::is_arithmetic<T>::value, Matrix>::type
+  Splat(T Val);
+  static Matrix Load(ByteAddressBuffer Res, uint StartOffset, uint Stride,
+                     bool ColMajor, uint Align = sizeof(ElementType));
+  static Matrix Load(RWByteAddressBuffer Res, uint StartOffset, uint Stride,
+                     bool ColMajor, uint Align = sizeof(ElementType));
 
   template <typename T>
-    requires ArithmeticScalar<T>
-  static Matrix Load(groupshared T Arr[], uint StartIdx, uint Stride,
-                     bool ColMajor);
+  static typename hlsl::enable_if<hlsl::is_arithmetic<T>::value, Matrix>::type
+  Load(/*groupshared*/ T Arr[], uint StartIdx, uint Stride, bool ColMajor);
 
-  void
-  Store(RWByteAddressBuffer Res, uint StartOffset, uint Stride, bool ColMajor,
-        uint Align = sizeof(ElementType));
+  void Store(RWByteAddressBuffer Res, uint StartOffset, uint Stride,
+             bool ColMajor, uint Align = sizeof(ElementType));
 
   template <typename T>
-    requires ArithmeticScalar<T>
-  void Store(groupshared T Arr[], uint StartIdx, uint Stride,
-             bool ColMajor);
+  typename hlsl::enable_if<hlsl::is_arithmetic<T>::value, void>::type
+  Store(/*groupshared*/ T Arr[], uint StartIdx, uint Stride, bool ColMajor);

```

Comment:
Why was this commented?

Reply:
We don't actually support `groupshared` as a function parameter qualifier for user-defined functions. We have some special handling for the interlocked intrinsics, and we'll likely need something similar to support this.